zeroize/src/lib.rs

```diff
@@ -360,6 +360,31 @@ impl Zeroize for String {
     fn zeroize(&mut self) {
         unsafe { self.as_bytes_mut() }.zeroize();
         debug_assert!(self.as_bytes().iter().all(|b| *b == 0));
+
+        // Zero the capacity of the `String` that is not initialized.
+        {
+            // Safety:
+            //
+            // This is safe, because `String` never allocates more than `isize::MAX` bytes.
+            let extra_capacity_start = unsafe { self.as_mut_ptr().add(self.len()) };
+            let extra_capacity_len = self.capacity().saturating_sub(self.len());
+
+            for i in 0..extra_capacity_len {
+                // Safety:
+                //
+                // This is safe, because `String` never allocates more than `isize::MAX` bytes.
+                let current_ptr = unsafe { extra_capacity_start.add(i) };
+                // Safety:
+                //
+                // `current_ptr` is valid, because it lies within the allocation of the `Vec`.
+                // It is also properly aligned, because we write a `u8`, which has an alignment of
+                // 1.
+                unsafe { ptr::write_volatile(current_ptr, 0) };
+            }
```

Comment:
This `for` loop is duplicated between the `Vec` and `String` serializing code. It also seems like the sort of thing which could be potentially unified with the existing [`volatile_set`](https://github.com/iqlusioninc/crates/blob/470678cc/zeroize/src/lib.rs#L459) method (converting the latter to use pointers instead of slices).

I think it'd be nice to have a single method as the place to focus on `memset`-like optimizations (although I can still see separating one that works on `u8` versus one that works on `T: Copy + Sized`)

Reply:
Looks good now, thanks!